[CASUtil] Implement CASIDFile

cachemeifyoucan opened the pull request:
Add CASIDFile implementation which is a YAML file capture both the CASID
and the CASDB instance information so the CAS object referred by the
CASID can be fetched from the information only from this file.

cachemeifyoucan commented:
This is a pitch for an improved file format for existing `CASID:` file.

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

dexonsmith commented on llvm/include/llvm/CAS/Utils.h:
```diff
@@ -37,6 +33,29 @@ Error walkFileTreeRecursively(
+
+/// CAS file
+/// A file on disk that represents an item in the CAS.
+class CASIDFile {
```
Feels like a big enough concept it could have its own header/source file.

dexonsmith commented on llvm/include/llvm/CAS/CASID.h:
```diff
@@ -28,12 +28,17 @@ class CASIDContext {
+  /// can be used together with schema identifier to retrive the same
+  /// CASIDContext.
+  virtual StringRef getHashInstanceIdentifier() const { return StringRef(); }
+
```
1. I don't think this string has anything to do with the "hash" so I don't think the name is right.
2. I think this is tied strongly to the CAS object storage — this is an identifier for loading CAS objects. (I think a better place/name would be `CASDB::getObjectStoreIdentifier()`.)
3. Nitpick: If it's optional, I think "none" should be modelled using an `None` rather than an empty string.

dexonsmith commented on llvm/tools/llvm-cas/llvm-cas.cpp:
```diff
@@ -33,6 +33,9 @@ static cl::list<std::string> PrefixMapPaths(
     "prefix-map",
     cl::desc("prefix map for file system ingestion, -prefix-map BEFORE=AFTER"));
 
+cl::opt<std::string> Output("o", cl::desc("CASID output file path"),
```
For `llvm-cas`, I wonder if default output should still be textual (as it was before), and this just changes where the output is written. That seems more convenient for tests which might want to use the ID on the command-line. Adding `-o` seems great, but shouldn't change the format.

I think a separate option should be added to change the format to this new format.  Something like `-emit-casfile`.

dexonsmith commented on llvm/tools/llvm-cas/llvm-cas.cpp:
```diff
@@ -514,6 +518,22 @@ int getCASIDForFile(CASDB &CAS, CASID ID, StringRef Path) {
+}
+
+int writeCASIDOutput(CASID ID) {
+  outs() << ID << "\n";
```
I don't think we should do both `outs()` and a file.

dexonsmith commented on llvm/tools/llvm-cas/llvm-cas.cpp:
```diff
@@ -514,6 +518,22 @@ int getCASIDForFile(CASDB &CAS, CASID ID, StringRef Path) {
+
+  ExitOnError ExitOnErr("llvm-cas: write CASID output file: ");
+  std::error_code EC;
+  raw_fd_ostream OS(Output, EC);
```
I suggest defaulting `Output` to `"-"` so that you can use `raw_fd_ostream` all the time (the constructor takes over `outs()` in that case).

dexonsmith commented on llvm/tools/llvm-cas/llvm-cas.cpp:
```diff
@@ -514,6 +518,22 @@ int getCASIDForFile(CASDB &CAS, CASID ID, StringRef Path) {
+  if (EC)
+    ExitOnErr(errorCodeToError(EC));
+
+  writeCASIDFile(OS, ID);
```
As mentioned elsewhere, I think whether the ID is streamed or `writeCASIDFile()` is called should depend on a separate command-line option.

dexonsmith commented on llvm/unittests/CAS/CASUtilsTest.cpp:
```diff
@@ -57,3 +57,22 @@ TEST(CASUtilsTest, walkFileTreeRecursively) {
+  auto Blob = cantFail(CAS->createBlob("blob"));
+
+  std::string FileStr;
+  raw_string_ostream SS(FileStr);
```
FYI, `SmallVector`/`SmallString` + `raw_svector_stream` has some advantages, since the latter is a `raw_pwrite_stream` and the former is not.

dexonsmith reviewed (commented):
I think this CAS-file abstraction is a really interesting idea.

As implemented, this will "leak" the path to the on-disk CAS, which `clang::CASOptions` has worked hard to hide so we can prove that it doesn't affect computation. Perniciously, if it's adopted by Clang for the `.o` files there, it'll actively be changing the output: exactly the problem that `CASOptions` was trying to design away!

That said, it seems useful to have a file that knows the path to the on-disk CAS.

Ideally, anyone creating this file would be at a higher layer, which has access to the original options used to create the CAS _a priori_, so they don't need to be leaked through. I haven't thought through the best design for that, but maybe others have some ideas.

cachemeifyoucan commented:
> I think this CAS-file abstraction is a really interesting idea.
> 
> As implemented, this will "leak" the path to the on-disk CAS, which `clang::CASOptions` has worked hard to hide so we can prove that it doesn't affect computation. Perniciously, if it's adopted by Clang for the `.o` files there, it'll actively be changing the output: exactly the problem that `CASOptions` was trying to design away!

Yes, this is just a pitch of idea to replace the Util function in the same file. Some cleanup is definitely needed. This is to create a placeholder to communicate a CAS object on real file system.

Note this is an abstract file so it doesn't matter what the file content. For all the tools that models the input/output, they should use the actual content from the CAS which this file points to.

dexonsmith commented:
> > I think this CAS-file abstraction is a really interesting idea.
> > As implemented, this will "leak" the path to the on-disk CAS, which `clang::CASOptions` has worked hard to hide so we can prove that it doesn't affect computation. Perniciously, if it's adopted by Clang for the `.o` files there, it'll actively be changing the output: exactly the problem that `CASOptions` was trying to design away!
> 
> Yes, this is just a pitch of idea to replace the Util function in the same file. Some cleanup is definitely needed. This is to create a placeholder to communicate a CAS object on real file system.

As long as this information is not leaked via CASDB, but passed in to something high-level, seems fine in principle.

> Note this is an abstract file so it doesn't matter what the file content. For all the tools that models the input/output, they should use the actual content from the CAS which this file points to.

I'm not entirely sure what you mean by "abstract" here. I believe the current architecture captures this file and includes it in the cached CAS tree. If you add logging to `cc1_main.cpp` to print out the CASID of the final "result" that is stored in the cache, I believe you'll find that, with your patch applied, the result will depend on the path to the CAS. This is because the function you're replacing affects the output of the "compute" phase in `-cc1` (if the phases are canonicalize, (check-cache, compute, store-cache), replay).

dexonsmith commented:
To be clear, I like this idea, but I'm concerned about the layering of leaking the CASPath through CASDB.

I think you could implement this without leaking the CASPath if you built it on top of https://github.com/apple/llvm-project/pull/4215 or similar.

akyrtzi commented:
@cachemeifyoucan could you clarify, is the idea that the `CASIDFile` embeds enough information for a client to be able to retrieve the data without needing to know what CASDB to use in advance? Or is the CASDB related information in `CASIDFile` only for verification that the client has the right CASDB opened?
If the latter, is it worth making sure that moving the CASDB to a different path should not fail loading the data from the same  `CASIDFile`?

dexonsmith commented:
> To be clear, I like this idea, but I'm concerned about the layering of leaking the CASPath through CASDB.
> 
> I think you could implement this without leaking the CASPath if you built it on top of #4215 or similar.

Note: I've landed that now. I think what you could do is:

- Change CompileJobCache to save the CASPath before it gets hidden.
- Pass CASPath as an argument to `writeCASIDBuffer()`

(no need to modify CASIDContext or CASDB)

cachemeifyoucan commented:
> @cachemeifyoucan could you clarify, is the idea that the `CASIDFile` embeds enough information for a client to be able to retrieve the data without needing to know what CASDB to use in advance? Or is the CASDB related information in `CASIDFile` only for verification that the client has the right CASDB opened?
> If the latter, is it worth making sure that moving the CASDB to a different path should not fail loading the data from the same `CASIDFile`?

The idea is the first one but I provided an interface for the latter if the tool makes the assumption about the expected location for CASDB. The main driver behind this file format is for a binary tool, like `nm` or `otool`, reading a CASID based CAS object file, it doesn't have the context for where the CASDB is. The use case for that might be a tool that reads multiple object files, expect them be in the same CASDB and what to avoid creating too many CASDB. The other solution will be we require the user of all those tools provided the same CASOption when reading the file.

@dexonsmith does have a point. We need to make sure this CASIDFile is just a OnDisk representation for a CAS Object and we need to layer it correct so the content of the file is never put into a CAS or being used to compute dependency. Maybe it should be written by CASOutputBackend. Let me think about this a bit more.

akyrtzi commented:
> The idea is the first one 

Ah ok, does that mean that `CASIDFile` should be recording enough info that a `CASOptions` object could be formed out of it?

cachemeifyoucan commented:
> Ah ok, does that mean that `CASIDFile` should be recording enough info that a `CASOptions` object could be formed out of it?

That is another way to look at it. My current implementation is to let the CASContext itself to provide enough information to recreate itself. The alternative is we can sink `CASOptions` into llvm and put that into the CASIDFile.

akyrtzi commented:
> we can sink CASOptions into llvm

That seems like general goodness to me.

cachemeifyoucan committed:
Sink CASOptions into LLVM

Move CASOptions to LLVM so other tools can share the same CASOptions to
initialize CAS.

cachemeifyoucan committed:
Switch CAS tools to use CASOptions

cachemeifyoucan committed:
Implement CASIDFile from CASOptions

Teach CASOptions to write a CASIDFile that can recreate the
CASOptions and decode the CASID to fetch the object.

cachemeifyoucan committed:
Teach llvm-cas to write CASIDFile

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

dexonsmith commented on llvm/lib/CAS/CASOptions.cpp:
```diff
@@ -92,3 +93,59 @@ void CASOptions::ensurePersistentCAS() {
+  raw_string_ostream SS(Output.CASID);
+  ID.print(SS);
+  // Use the CASPath from the cache as the frozen CASOption have the path wiped.
+  Output.CASPath = Cache.Config.CASPath;
```
This still feels like an information leak to me. Now anyone that wants to can sniff out the CAS configuration by writing to and reading from a `SmallVector` / `raw_svector_ostream`, and there's no layering to prevent this output file from being cached by something that wants to be independent of the CAS location.

Instead, this should use `CASPath`, with a comment "not using Cache.Config.CASPath, since that would leak the hidden configuration", or take an optional `CASConfiguration*=nullptr` as an argument to pass in settings. Th elatter, I'm thinking:
```
void CASOptions::writeCASIDFile(raw_ostream &OS, const CASID &ID, const CASConfiguration *ConfigToSerialize) const {
  if (!ConfigToSerialize)
    ConfigToSerialize = this;
```

For clang, you can just save this up a level, and pass it in as a parameter. Here's the current code. Note the TODO:
```
  // Hide the CAS configuration, canonicalizing it to keep the path to the
  // CAS from leaking to the compile job, where it might affecting its
  // output (e.g., in a diagnostic).
  //
  // TODO: Extract CASOptions.Path first if we need it later since it'll
  // disappear here.
  CAS = Invocation.getCASOpts().getOrCreateCASAndHideConfig(Diags);
```
You can replace it with:
```
  CASOptionsConfig = Invocation.getCASOpts();
  CAS = Invocation.getCASOpts().getOrCreateCASAndHideConfig(Diags);
```
adding `CASConfiguration` field to CompileJobCache, and then pass in the saved config as an extra argument when writing the file, or:
```
  CAS = Invocation.getCASOpts().getOrCreateCAS(Diags);
  CASOpts = Invocation.getCASOpts();
  Invocation.getCASOpts().hideConfig(); // new function to implement.
```
adding `CASOptions` field to CompileJobCache.

dexonsmith replied:
I wonder, maybe only `CASConfiguration` should be sunk down into LLVM, and this could just be a free function that takes `const CASConfiguration&` by argument. Then there's no need to override it.
- Clang can save the config in the CompileJobCache and pass it in when writing out the file in `CompileJobCache::finishComputedResult()`
- `CASOptions` is free to reference clang-specific things like `DiagnosticsEngine`

cachemeifyoucan replied:
Technically, I didn't change CASOption API at all so what is leaking is what is currently leaking. The interface needed to write CASID file needs to leak the path as the target audience is not clang but binary tools that are not running inside a build system like `nm`. Here my current thinking. 

* Sink CASConfiguration into llvm, and it is ok to be leaked to CASPath and it is part of the API.
* CASOption is clang specific. It is not inheriting from CASConfiguration, but use that as a private field, so CASOption is created from CASConfiguration but nothing can get the CASPath out of the CASOption. If clang wants to write a CASID file, the CacheJobAction needs to keep track of the CASConfiguration used to create CASOption and use that.

dexonsmith replied:
> Technically, I didn't change CASOption API at all so what is leaking is what is currently leaking.

I disagree. Technically, you changed the CASOptions API and started leaking the hidden configuration. Previously if someone received `CASOptions`, they could not accidentally depend on a hidden config. Now they can! E.g.:
```
auto doSomething(const CASOptions &HiddenOpts, OtherArgs) {
  CASID Result1 = compute1(HiddenOpts, OtherArgs); // Is Result1 tainted? Hopefully not...
  SmallVector<char> Result2;
  raw_svector_ostream OS(Result2);
  HiddenOpts.writeCASIDFile(OS, Result1); // Result2 has been tainted (accidental leak)
  auto Result3 = compute3(HiddenOpts, Result2); // Result3 has been tainted (accidental leak)

  // Direct leak: read the config back. Depending on how deeply this logic is buried,
  // this could also be an accidental leak.
  CASOptions FoundOpts;
  FoundOpts.createFromCASIDFile(MemoryBufferRef(Result2));

  // ...
```

> The interface needed to write CASID file needs to leak the path as the target audience is not clang but binary tools that are not running inside a build system like nm.

I understand that there is a practical use case with a target audience that does not include clang caching. The point of contention is on how to architect it such that it doesn't have accidental side effects on clang caching.

I think it's important that the architecture for caching is hermetic (leak-free). It needs to be clear and obvious what the inputs are. If there's something we don't want to affect results, then we cannot let it be observable, or we cannot trust that the computations are pure. If we can't trust the computations to be pure, it's not sound to cache them.

> Here my current thinking. [snip]

Yup, I think that matches my thinking. To be clear, if `CASOptions` has hidden its configuration, then there should be no way to observe that configuration (and have it become an accidental input). But passing the original (saved) `CASConfiguration` *as an input* to some other computation (such a function that creates a file from a CASID+CASConfig) sounds fine. It'll affect that output, but since the config is an explicit input to the computation, that seems fine.

However, note that anything checking for reproducible builds will see that the environment (path to the CAS) has affected the object file. There's a tradeoff there.
